core/src/main/scala/kafka/tools/ConsoleProducer.scala

```diff
@@ -264,38 +282,113 @@ object ConsoleProducer {
-        case (line, true) =>
-          line.indexOf(keySeparator) match {
-            case -1 =>
-              if (ignoreError) new ProducerRecord(topic, line.getBytes(StandardCharsets.UTF_8))
-              else throw new KafkaException(s"No key found on line $lineNumber: $line")
-            case n =>
-              val value = (if (n + keySeparator.size > line.size) "" else line.substring(n + keySeparator.size)).getBytes(StandardCharsets.UTF_8)
-              new ProducerRecord(topic, line.substring(0, n).getBytes(StandardCharsets.UTF_8), value)
-          }
-        case (line, false) =>
-          new ProducerRecord(topic, line.getBytes(StandardCharsets.UTF_8))
+      if (printPrompt) print(">")
+      val line = reader.readLine()
+      try{
+        (line, parseKey, parseHeader) match {
+          case (null, _, _) => null
+          case (line, true, true) => parseHeaderKeyValue(line)
+          case (line, true, false) => parseKeyValue(line)
+          case (line, false, true) => parseHeaderValue(line)
+          case (line, false, false) => parseValue(line)
+        }
+      } catch {
+        case _: Throwable => onMatchError(line)
+      }
+    }
+
+    private def parseHeaderKeyValue(line: String): ProducerRecord[Array[Byte],Array[Byte]] = {
+      val headerIndex = line.indexOf(headersDelimiter)
+      val headers = line.substring(0, headerIndex)
+      val keyValue = line.substring(headerIndex + 1)
+      val keyIndex = keyValue.indexOf(keySeparator)
+      val key = keyValue.substring(0, keyIndex)
+      val value = keyValue.substring(keyIndex + 1)
+      new ProducerRecord(topic, null, null, key.getBytes(StandardCharsets.UTF_8), value.getBytes(StandardCharsets.UTF_8), mapHeaders(headers))
+    }
+
+    private def parseKeyValue(line: String): ProducerRecord[Array[Byte],Array[Byte]] = {
+      val keyIndex = line.indexOf(keySeparator)
+      val key = line.substring(0, keyIndex)
+      val value = line.substring(keyIndex + 1)
+      new ProducerRecord(topic, key.getBytes(StandardCharsets.UTF_8), value.getBytes(StandardCharsets.UTF_8))
+    }
+
+    private def parseHeaderValue(line: String): ProducerRecord[Array[Byte],Array[Byte]] = {
+      val headerIndex = line.indexOf(headersDelimiter)
+      val headers = line.substring(0, headerIndex)
+      val value = line.substring(headerIndex + 1)
+      new ProducerRecord(topic, null, null, null, value.getBytes(StandardCharsets.UTF_8), mapHeaders(headers))
+    }
+
+    private def parseValue(line: String): ProducerRecord[Array[Byte],Array[Byte]] = {
+      new ProducerRecord(topic, line.getBytes(StandardCharsets.UTF_8))
+    }
+
+    private def onMatchError(line: String): ProducerRecord[Array[Byte], Array[Byte]] = {
+      if (ignoreError) {
+        new ProducerRecord(topic, line.getBytes(StandardCharsets.UTF_8))
+      } else {
+        throw new KafkaException("Could not parse line " + lineNumber + ", most likely line does not match pattern: " + illustratePattern())
```

Comment:
We can use string interpolation here: `KafkaException(s"Could not parse line $lineNumber ...` rather than concatenation